packages/javascript-decorators/src/memoize.ts

```diff
@@ -0,0 +1,74 @@
+export interface MemoizeMap<T, U> {
+  get(key: T): U;
+  has(key: T): boolean;
+  set(key: T, value: U): MemoizeMap<T, U>;
+}
+
+type Resolver<T extends Function> = T extends (...args: infer T) => any
+  ? T
+  : Function;
+
+export const MAX_MAP_ENTRIES = 50;
+
+export default function memoize<F extends Function>(
+  resolver?: Resolver<F>,
+): MethodDecorator {
+  return function<F>(
+    _target: Object,
+    _propertyKey: string | symbol,
+    descriptor: TypedPropertyDescriptor<F>,
+  ) {
+    const method: F | undefined = descriptor.value;
+
+    if (!method || !(method instanceof Function)) {
+      return descriptor;
+    }
+
+    // @ts-ignore
+    descriptor.value = memoized<F>(method, resolver);
```

Comment:
The error here without `ts-ignore`
```
Type 'F' does not satisfy the constraint 'Function'.
```

if I remove `<F>`, I get an error on `resolver`
```
Argument of type 'Resolver<F> | undefined' is not assignable to parameter of type 'Resolver<F & Function> | undefined'.
  Type 'Resolver<F>' is not assignable to type 'Resolver<F & Function> | undefined'.
    Type 'Function | unknown[]' is not assignable to type 'Resolver<F & Function> | undefined'.
      Type 'Function' is not assignable to type 'Resolver<F & Function>'.
        Type 'Function' is not assignable to type 'Resolver<F & Function>'.
          Type 'Resolver<F>' is not assignable to type 'Resolver<F & Function>'.
```

Reply:
nvm I figure it out